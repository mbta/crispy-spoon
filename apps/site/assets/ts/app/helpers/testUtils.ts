/* eslint-disable */
import toJson, { Json } from "enzyme-to-json";
import { ReactWrapper } from "enzyme";
import {
  IndexedRouteStop,
  StopTree
} from "../../schedule/components/__schedule";

export const testRouteStopListFromStopTree = (
  tree: StopTree
): IndexedRouteStop[] =>
  Object.values(tree.byId).map((node, index) => ({
    ...node.value,
    routeIndex: index
  }));

export const createReactRoot = (): void => {
  document.body.innerHTML =
    '<div><div id="react-root"><div id="test"></div></div></div>';
};

export const enzymeToJsonWithoutProps = (wrapper: ReactWrapper<any>): Json =>
  toJson(wrapper, {
    noKey: false,
    map: info => {
      if (!info.props) {
        return info;
      }

      // @ts-ignore is a <style></style> declaration
      if (info.node.type === "style") {
        return info;
      }
      // @ts-ignore is an element in a <noscript></noscript> declaration
      if (info.node.nodeType == undefined) {
        // @ts-ignore
        return info;
      }
      //@ts-ignore is a React component (not HTML element)
      if (info.node.nodeType.toString().includes("function")) {
        // @ts-ignore
        info.props = undefined;
      }

      return info;
    }
  });

export const toJsonEnzyme = (wrapper: ReactWrapper<any>): Json =>
  toJson(wrapper, {
    noKey: false
  });
