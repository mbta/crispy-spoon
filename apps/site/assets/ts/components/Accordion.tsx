import React, { ReactElement, useState } from "react";

// A simple accordion implementation, based on site_web/templates/partial/_accordion_ui
// This attempts to mimic the usage of <ExpandableBlock />, where the contents rerender with expanded state
interface Props {
  id: string;
  title: {
    collapsed: string;
    expanded: string;
  };
  children: ReactElement<HTMLElement>;
}

export const AccordionNoJS = ({
  id,
  children
}: {
  id: string;
  children: ReactElement<HTMLElement>;
}): ReactElement<HTMLElement> => (
  <div className="c-accordion-ui">
    <div
      className="c-accordion-ui__target"
      id={`${id}-section`}
      role="region"
      aria-labelledby={`${id}-title`}
    >
      <div className="c-accordion-ui__content">{children}</div>
    </div>
  </div>
);

const Accordion = (props: Props): ReactElement<HTMLElement> => {
  const { id, title, children } = props;
  const [isExpanded, toggleExpanded] = useState(false);
  const onClick = (): void => toggleExpanded(expanded => !expanded);
  return (
    <div className="c-accordion-ui">
      <div className="panel">
        <div className="c-accordion-ui__heading" role="heading" aria-level={3}>
          <button
            className={`c-accordion-ui__trigger ${!isExpanded && "collapsed"}`}
            aria-expanded={isExpanded}
            aria-controls={`${id}-section`}
            onClick={onClick}
            type="button"
          >
            <span className="c-accordion-ui__title" id={`${id}-title`}>
              {isExpanded ? title.expanded : title.collapsed}
            </span>
            <div className="c-accordion-ui__indicator">
              <span className="c-indicator__content c-indicator__content--angle" />
            </div>
          </button>
        </div>
        <div
          className="c-accordion-ui__target js-focus-on-expand"
          id={`${id}-section`}
          role="region"
          aria-labelledby={`${id}-title`}
        >
          {isExpanded && (
            <div className="c-accordion-ui__content">{children}</div>
          )}
        </div>
      </div>
    </div>
  );
};

export default Accordion;
