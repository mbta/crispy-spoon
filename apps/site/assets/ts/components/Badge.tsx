import React from "react";

const Badge = ({
  text,
  contextText,
  bgClass
}: {
  text: string;
  contextText?: string;
  bgClass?: string;
}): JSX.Element => {
  return (
    <div
      className={`${bgClass ||
        "u-error-background"} font-weight-bold ps-8 pe-8 fs-14`}
      style={{ borderRadius: "0.75rem" }}
    >
      {/* The purpose of this block is to have invisble text for screen readers */}
      {contextText && <span className="sr-only">{contextText}</span>}
      {text}
    </div>
  );
};

export default Badge;
