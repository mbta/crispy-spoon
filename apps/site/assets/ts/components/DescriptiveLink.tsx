import React from "react";
import renderFa from "../helpers/render-fa";

interface DescriptiveLinkProps {
  href: string;
  title: string;
  bodyTextOrHtml: string; // might be HTML from the CMS
}

/* Since the props can be populated by running CMS content through
`Site.ContentRewriter.rewrite` this might end up using an empty string in place
of null values. So we can't do our usual undefined/null checks. This helper
function will substitute. */
const hasContent = (text: string): boolean => text !== "";

const DescriptiveLink = ({
  href,
  title,
  bodyTextOrHtml
}: DescriptiveLinkProps): JSX.Element | null => {
  const titleContent = hasContent(title) && (
    <div className="c-descriptive-link__title">{title}</div>
  );
  const bodyContent = hasContent(bodyTextOrHtml) && (
    // eslint-disable-next-line react/no-danger
    <div dangerouslySetInnerHTML={{ __html: bodyTextOrHtml }} />
  );

  if (!titleContent && !bodyContent) {
    return null;
  }

  return (
    <a className="c-descriptive-link" href={href}>
      <div className="c-descriptive-link__text">
        {titleContent}
        {bodyContent}
      </div>
      <div className="c-descriptive-link__caret-wrapper">
        {renderFa("c-descriptive-link__caret", "fa-solid fa-angle-right")}
      </div>
    </a>
  );
};

export default DescriptiveLink;
