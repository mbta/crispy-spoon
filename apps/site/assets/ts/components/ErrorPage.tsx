import React from "react";
import { FallbackRender } from "@sentry/react";

// Adapted from the templates in SiteWeb.ErrorView
// To be used in an ErrorBoundary
const ErrorPage: FallbackRender = errorData => {
  // eslint-disable-next-line no-console
  console.error("from Sentry", JSON.stringify(errorData));
  return (
    <div className="container error-page">
      <div className="row">
        <div className="col-md-6">
          <h1 className="mt-1">Oh no! We&#39;re experiencing delays.</h1>
          <p className="error-paragraph u-bold">
            Something went wrong on our end.
          </p>
          <div className="error-links">
            Go to{" "}
            <a className="c-call-to-action" href="/">
              {`the MBTA's home page`}
            </a>{" "}
            or{" "}
            <a className="c-call-to-action" href="/customer-support">
              report an issue
            </a>
            .
          </div>
        </div>
        <div className="col-md-6">
          <img alt="" src="/images/error-bus.gif" className="w-100" />
        </div>
      </div>
    </div>
  );
};

export default ErrorPage;
