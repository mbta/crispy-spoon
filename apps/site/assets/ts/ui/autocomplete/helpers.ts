import { omit } from "lodash";
import { OnSubmitParams } from "@algolia/autocomplete-core";
import { OnStateChangeProps } from "@algolia/autocomplete-js";
import {
  ContentItem,
  Item,
  LocationItem,
  PopularItem,
  RouteItem,
  StopItem
} from "./__autocomplete";
import { isLGDown } from "../../helpers/media-breakpoints";

export function isStopItem(x: Item): x is StopItem {
  return Object.keys(x).includes("stop");
}

export function isRouteItem(x: Item): x is RouteItem {
  return Object.keys(x).includes("route");
}

export function isContentItem(x: Item): x is ContentItem {
  return Object.keys(x).includes("_content_type");
}

export const getTitleAttribute = (item: Item): string[] => {
  if (isStopItem(item)) {
    return ["stop", "name"];
  }
  if (isRouteItem(item)) {
    return ["route", "name"];
  }

  // @ts-ignore
  // eslint-disable-next-line no-underscore-dangle
  return Object.keys(item._highlightResult);
};

const navStateChange: (props: OnStateChangeProps<Item>) => void = ({
  state
}) => {
  // grey out the page and disable scrolling when search is open
  if (isLGDown()) {
    if (state.isOpen) {
      document.documentElement.dataset.navOpen = "true";
    } else {
      delete document.documentElement.dataset.navOpen;
    }
  }
};

export const STATE_CHANGE_HANDLERS: Record<
  string,
  (props: OnStateChangeProps<Item>) => void
> = {
  nav: navStateChange
};

export const SUBMIT_HANDLERS: Record<
  string,
  (props: OnSubmitParams<Item>) => void
> = {
  to_search_page({ state }) {
    window.Turbolinks.visit(`/search?query=${state.query}`);
  }
};

export type WithUrls<T> = T & { urls: Record<string, string> };

// The backend returns all possible URLs, use urlType to get the desired one
export const itemWithUrl = (
  initialItem: WithUrls<LocationItem | PopularItem>,
  urlType: string
): Omit<typeof initialItem, "urls"> => {
  const item = omit(initialItem, "urls");
  item.url = initialItem.urls[urlType];
  return item;
};
