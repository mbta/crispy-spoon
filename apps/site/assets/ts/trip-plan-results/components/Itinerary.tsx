import React, { ReactElement, useEffect } from "react";
import Accordion, { AccordionNoJS } from "../../components/Accordion";
import { Itinerary } from "./TripPlannerResults";
import Map from "../../leaflet/components/Map";
import {
  addAlertItemEventHandlers,
  removeAlertItemEventHandlers
} from "../../../js/alert-item";

/*
The following ignored functions are vanilla JS event handlers
that are for HTML that gets injected into itinerary rows.
Ideally, all the HTML of this component would be built in React
which would make it easy to test in this file. Since that HTML is
server rendered and injected, it is difficult to test from jest.
It is being ignored now to prevent a codecov ding.
*/

/* istanbul ignore next */
const toggleTextContent = (e: Event): void => {
  if ((e.target! as HTMLElement).textContent!.trim() === "(view alert)") {
    (e.target! as HTMLElement).textContent = "(hide alert)";
  } else {
    (e.target! as HTMLElement).textContent = "(view alert)";
  }
};

/* istanbul ignore next */
export const addToggleAlertHandlers = (): void => {
  Array.from(document.querySelectorAll(`.js-trip-plan-alert-toggle`)).forEach(
    elem => {
      elem.addEventListener("click", toggleTextContent);
    }
  );
};

/* istanbul ignore next */
export const removeToggleAlertHandlers = (): void => {
  Array.from(document.querySelectorAll(`.js-trip-plan-alert-toggle`)).forEach(
    elem => {
      elem.removeEventListener("click", toggleTextContent);
    }
  );
};

interface Props {
  itinerary: Itinerary;
}

const ItineraryBody = (itinerary: Itinerary): ReactElement<HTMLElement> => {
  useEffect(() => {
    addAlertItemEventHandlers();
    addToggleAlertHandlers();
    /* istanbul ignore next */
    return () => {
      removeAlertItemEventHandlers();
      removeToggleAlertHandlers();
    };
  }, []);

  return (
    <div className="m-trip-plan-results__itinerary-body">
      <div className="trip-plan-map map">
        <Map mapData={itinerary.map} boundsByMarkers />
      </div>
      <div
        dangerouslySetInnerHTML={{ __html: itinerary.html }} // eslint-disable-line react/no-danger
      />
    </div>
  );
};

const ItineraryAccordion = ({
  itinerary
}: Props): ReactElement<HTMLElement> => (
  <div className="m-trip-plan-results__itinerary">
    <div className="m-trip-plan-results__itinerary-header">
      <div className="m-trip-plan-results__itinerary-header-content">
        <div
          className="m-trip-plan-results__itinerary-summary"
          dangerouslySetInnerHTML={{ __html: itinerary.tab_html }} // eslint-disable-line react/no-danger
        />
        <div
          className="m-trip-plan-results__itinerary-accessible"
          dangerouslySetInnerHTML={{ __html: itinerary.access_html }} // eslint-disable-line react/no-danger
        />
      </div>
      <div
        className="m-trip-plan-results__itinerary-fares"
        dangerouslySetInnerHTML={{ __html: itinerary.fares_estimate_html }} // eslint-disable-line react/no-danger
      />
    </div>
    <Accordion
      id={`itinerary-${itinerary.id}`}
      title={{
        collapsed: "Show map and trip details",
        expanded: "Hide map and trip details"
      }}
    >
      <ItineraryBody {...itinerary} />
    </Accordion>
    <noscript>
      <AccordionNoJS id={`itinerary-${itinerary.id}`}>
        <ItineraryBody {...itinerary} />
      </AccordionNoJS>
    </noscript>

    <Accordion
      id={`itinerary-${itinerary.id}-fare-calculator`}
      title={{
        collapsed: "Show fare calculator",
        expanded: "Hide fare calculator"
      }}
    >
      <div
        dangerouslySetInnerHTML={{ __html: itinerary.fare_calculator_html }} // eslint-disable-line react/no-danger
      />
    </Accordion>
    <noscript>
      <AccordionNoJS id={`itinerary-${itinerary.id}-fare-calculator`}>
        <div
          dangerouslySetInnerHTML={{ __html: itinerary.fare_calculator_html }} // eslint-disable-line react/no-danger
        />
      </AccordionNoJS>
    </noscript>
  </div>
);

export default ItineraryAccordion;
