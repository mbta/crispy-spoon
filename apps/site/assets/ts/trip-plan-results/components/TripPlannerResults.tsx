import React, { ReactElement } from "react";
import { MapData } from "../../leaflet/components/__mapdata";
import Itinerary from "./Itinerary";

export interface Itinerary {
  id: number;
  html: string;
  // eslint-disable-next-line
  map: MapData;
  tab_html: string;
  access_html: string;
  fares_estimate_html: string;
  fare_calculator_html: string;
}

interface Props {
  // eslint-disable-next-line
  itineraryData: Itinerary[];
}

const TripPlannerResults = ({
  itineraryData
}: Props): ReactElement<HTMLElement> => (
  <>
    {itineraryData.map(itinerary => (
      <Itinerary key={itinerary.id} itinerary={itinerary} />
    ))}
  </>
);

export default TripPlannerResults;
