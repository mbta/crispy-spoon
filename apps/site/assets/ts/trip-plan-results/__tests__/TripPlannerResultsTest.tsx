import React from "react";
import renderer, { act } from "react-test-renderer";
import { createReactRoot } from "../../app/helpers/testUtils";
import TripPlannerResults from "../components/TripPlannerResults";
import Accordion, { AccordionNoJS } from "../../components/Accordion";
import { TileServerUrl } from "../../leaflet/components/__mapdata";

const html = "<div>Lots of content about the itinerary</div>";
const tab_html = "<div>HTML for the accordion button</div>";
const access_html = "<div>Accessible</div>";
const fares_estimate_html = "<div>HTML content for fares</div>";
const fare_calculator_html = "<div>HTML content for fare calculator</div>";
const tile_server_url: TileServerUrl = "";

const positions: [number, number][] = [
  [42.3564, -71.06242],
  [42.35628, -71.06226]
];
const map = {
  zoom: null,
  markers: [],
  tile_server_url,
  default_center: { longitude: 0, latitude: 0 },
  height: 400,
  width: 400,
  polylines: [
    {
      color: "#428608",
      "dotted?": false,
      id: "",
      positions,
      weight: 5
    }
  ]
};

it("it renders", () => {
  createReactRoot();
  const tree = renderer.create(
    <TripPlannerResults
      itineraryData={[
        {
          html,
          access_html,
          fares_estimate_html,
          tab_html,
          fare_calculator_html,
          id: 1,
          map
        }
      ]}
    />
  );
  tree.update(
    <TripPlannerResults
      itineraryData={[
        {
          html,
          access_html,
          fares_estimate_html,
          tab_html,
          fare_calculator_html,
          id: 1,
          map
        }
      ]}
    />
  );
  expect(tree.toJSON()).toMatchSnapshot();
});

it("it renders ItineraryBody when clicking to expand", () => {
  createReactRoot();
  const tree = renderer.create(
    <TripPlannerResults
      itineraryData={[
        {
          html,
          access_html,
          fares_estimate_html,
          tab_html,
          fare_calculator_html,
          id: 1,
          map
        }
      ]}
    />
  );

  const button = tree.root.findAllByType("button")[0];

  act(button.props.onClick);

  expect(tree.root.findAllByType(Accordion).length).toBe(2);
  expect(tree.root.findAllByType(AccordionNoJS).length).toBe(2);

  expect(
    tree.root.findAllByProps({
      html: "<div>Lots of content about the itinerary</div>"
    }).length
  ).toBe(2);
});
