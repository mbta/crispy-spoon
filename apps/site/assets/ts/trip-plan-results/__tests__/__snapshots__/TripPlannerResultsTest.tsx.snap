// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders 1`] = `
<div
  className="m-trip-plan-results__itinerary"
>
  <div
    className="m-trip-plan-results__itinerary-header"
  >
    <div
      className="m-trip-plan-results__itinerary-header-content"
    >
      <div
        className="m-trip-plan-results__itinerary-summary"
        dangerouslySetInnerHTML={
          Object {
            "__html": "<div>HTML for the accordion button</div>",
          }
        }
      />
      <div
        className="m-trip-plan-results__itinerary-accessible"
        dangerouslySetInnerHTML={
          Object {
            "__html": "<div>Accessible</div>",
          }
        }
      />
    </div>
    <div
      className="m-trip-plan-results__itinerary-fares"
      dangerouslySetInnerHTML={
        Object {
          "__html": "<div>HTML content for fares</div>",
        }
      }
    />
  </div>
  <div
    className="c-accordion-ui"
  >
    <div
      aria-level={3}
      className="panel"
      role="heading"
    >
      <div
        className="c-accordion-ui__heading"
      >
        <button
          aria-controls="itinerary-1-section"
          aria-expanded={false}
          className="c-accordion-ui__trigger collapsed"
          onClick={[Function]}
          type="button"
        >
          <span
            className="c-accordion-ui__title"
            id="itinerary-1-title"
          >
            Show map and trip details
          </span>
          <div
            className="c-accordion-ui__indicator"
          >
            <span
              className="c-indicator__content c-indicator__content--angle"
            />
          </div>
        </button>
      </div>
      <div
        aria-labelledby="itinerary-1-title"
        className="c-accordion-ui__target js-focus-on-expand"
        id="itinerary-1-section"
        role="region"
      />
    </div>
  </div>
  <div
    className="c-accordion-ui"
  >
    <div
      aria-level={3}
      className="panel"
      role="heading"
    >
      <div
        className="c-accordion-ui__heading"
      >
        <button
          aria-controls="itinerary-1-fare-calculator-section"
          aria-expanded={false}
          className="c-accordion-ui__trigger collapsed"
          onClick={[Function]}
          type="button"
        >
          <span
            className="c-accordion-ui__title"
            id="itinerary-1-fare-calculator-title"
          >
            Show fare calculator
          </span>
          <div
            className="c-accordion-ui__indicator"
          >
            <span
              className="c-indicator__content c-indicator__content--angle"
            />
          </div>
        </button>
      </div>
      <div
        aria-labelledby="itinerary-1-fare-calculator-title"
        className="c-accordion-ui__target js-focus-on-expand"
        id="itinerary-1-fare-calculator-section"
        role="region"
      />
    </div>
  </div>
</div>
`;
